shader/fine.wgsl

```diff
@@ -260,88 +269,315 @@ fn fill_path_ms(fill: CmdFill, wg_id: vec2<u32>, local_id: vec2<u32>) -> array<f
+    packed_w += (packed_w - 0x808080u) << 8u;
+    packed_w += (packed_w - 0x8080u) << 16u;
+    var packed_y = atomicLoad(&sh_winding_y[local_id.y >> 2u]);
+    packed_y += (packed_y - 0x808080u) << 8u;
+    packed_y += (packed_y - 0x8080u) << 16u;
+    var wind_y = (packed_y >> ((local_id.y & 3u) << 3u)) - 0x80u;
+    if (local_id.y & 3u) == 3u && local_id.x == 0u {
+        let prefix_y = wind_y;
+        atomicStore(&sh_winding_y_prefix[local_id.y >> 2u], prefix_y);
     }
+    let prefix_x = ((packed_w >> 24u) - 0x80u) * 0x1010101u;
+    // reuse sh_winding to store prefix as well
+    atomicStore(&sh_winding[major], prefix_x);
     workgroupBarrier();
-    var wind_y = (packed_y >> ((local_id.y & 7u) << 2u)) - 8u;
-    if local_id.y >= 8u {
-        wind_y += (atomicLoad(&sh_winding_y[0]) >> 28u) - 8u;
+    for (var i = (major & ~3u); i < major; i++) {
+        packed_w += atomicLoad(&sh_winding[i]);
+    }
+    for (var i = 0u; i < local_id.y >> 2u; i++) {
+        wind_y += atomicLoad(&sh_winding_y_prefix[i]);
     }
 
     for (var i = 0u; i < PIXELS_PER_THREAD; i++) {
         let pix_ix = th_ix * PIXELS_PER_THREAD + i;
-        let minor = pix_ix & 7u;
-        //let nonzero = ((packed_w >> (minor << 2u)) & 0xfu) != u32(8 + backdrop);
-        // TODO: math might be off here
-        let expected_zero = (((packed_w >> (minor * 4u)) + wind_y) & 0xfu) - u32(fill.backdrop);
-        if expected_zero >= 16u {
+        let minor = i; // assumes PIXELS_PER_THREAD == 4
+        let expected_zero = (((packed_w >> (minor * 8u)) + wind_y) & 0xffu) - u32(fill.backdrop);
+        if expected_zero >= 256u {
             area[i] = 1.0;
         } else {
 #ifdef msaa8
-            let samples = atomicLoad(&sh_samples[pix_ix]);
-            let xored = (expected_zero * 0x11111111u) ^ samples;
-            // Each 4-bit nibble in xored is 0 for winding = 0, nonzero otherwise
-            let xored2 = xored | (xored * 2u);
-            let xored4 = xored2 | (xored2 * 4u);
-            area[i] = f32(countOneBits(xored4 & 0x88888888u)) * 0.125;
-#endif
-#ifdef msaa16
             let samples0 = atomicLoad(&sh_samples[pix_ix * 2u]);
             let samples1 = atomicLoad(&sh_samples[pix_ix * 2u + 1u]);
-            let xored0 = (expected_zero * 0x11111111u) ^ samples0;
+            let xored0 = (expected_zero * 0x1010101u) ^ samples0;
             let xored0_2 = xored0 | (xored0 * 2u);
-            let xored1 = (expected_zero * 0x11111111u) ^ samples1;
+            let xored1 = (expected_zero * 0x1010101u) ^ samples1;
             let xored1_2 = xored1 | (xored1 >> 1u);
             let xored2 = (xored0_2 & 0xAAAAAAAAu) | (xored1_2 & 0x55555555u);
             let xored4 = xored2 | (xored2 * 4u);
-            area[i] = f32(countOneBits(xored4 & 0xCCCCCCCCu)) * 0.0625;
+            let xored16 = xored4 | (xored4 * 16u);
+            area[i] = f32(countOneBits(xored16 & 0xC0C0C0C0u)) * 0.125;
 #endif
+#ifdef msaa16
```

Comment:
Please try to document your invariants around the SWAR math as best as you can. Perhaps even add a note that refers the reader to the even-odd version as it's easier to understand.